modules/core/lib/Auth/Process/SubjectID.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+declare(strict_types=1);
+
+namespace SimpleSAML\Module\core\Auth\Process;
+
+use Exception;
+use SAML2\Constants;
+use SAML2\XML\saml\NameID;
+use SimpleSAML\Assert\Assert;
+use SimpleSAML\Auth;
+
+/**
+ * Filter to generate the subject ID attribute.
+ *
+ * See: http://docs.oasis-open.org/security/saml-subject-id-attr/v1.0/csprd01/saml-subject-id-attr-v1.0-csprd01.html
+ *
+ * By default, this filter will generate the ID based on the UserID of the current user.
+ * This is generated from the attribute configured in 'identifyingAttribute' in the
+ * authproc-configuration.
+ *
+ * Example - generate from attribute:
+ * <code>
+ * 'authproc' => [
+ *   50 => [
+ *       'core:SubjectID',
+ *       'identifyingAttribute' => 'uid',
+ *       'scope' => 'example.org',
+ *   ]
+ * ]
+ * </code>
+ *
+ * @package SimpleSAMLphp
+ */
+class SubjectID extends Auth\ProcessingFilter
+{
+    /**
+     * The attribute we should generate the subject id from.
+     *
+     * @var string
+     */
+    private string $identifyingAttribute;
+
+    /**
+     * The scope to use for this attribute.
+     *
+     * @var string
+     */
+    private string $scope;
+
+    /**
+     * Initialize this filter.
+     *
+     * @param array &$config  Configuration information about this filter.
+     * @param mixed $reserved  For future use.
+     */
+    public function __construct(array &$config, $reserved)
+    {
+        parent::__construct($config, $reserved);
+
+        Assert::keyExists($config, 'identifyingAttribute', "Missing mandatory 'identifyingAttribute' config setting.");
+        Assert::keyExists($config, 'scope', "Missing mandatory 'scope' config setting.");
+        Assert::stringNotEmpty($config['identifyingAttribute']);
+        Assert::regex(
+            $config['scope'],
+            '/^[a-zA-Z0-9.-]+$/i',
```

Comment:
As above. Makes one wonder if we should factor this duplicated code out to a shared place. And/or combine the two attributes into a single AuthProc that you call with the type of identifier you want.

Reply:
Maybe have PairwiseID extend SubjectID? Other than that, I don't think it's a good idea to merge these two classes..
I have to give this some thought, but Jaime and I agree on the fact that they should be separate filters..

Reply:
Alright. There is some duplication also with the regexps being defined in two places etc.